src/state.rs

```diff
@@ -259,6 +282,175 @@ impl DemandProxyState {
+        // if/when we support that, this function may need to move to get access to the necessary metadata.
+        // Randomly pick an IP
+        // TODO: do this more efficiently, and not just randomly
+        if let Some(ip) = dst_workload.workload_ips.choose(&mut rand::thread_rng()) {
+            return Ok(*ip);
+        }
+        if dst_workload.hostname.is_empty() {
+            debug!(
+                "workload {} has no suitable workload IPs for routing",
+                dst_workload.name
+            );
+            return Err(Error::NoValidDestination(Box::new(dst_workload.clone())));
+        }
+        let ip = self
+            .load_balance_for_hostname(dst_workload, src_workload, metrics)
+            .await?;
+        Ok(ip)
+    }
+
+    async fn load_balance_for_hostname(
+        &self,
+        workload: &Workload,
+        src_workload: &Workload,
+        metrics: Arc<proxy::Metrics>,
+    ) -> Result<IpAddr, Error> {
+        let mut state: DemandProxyState = self.clone();
+        let labels = OnDemandDnsLabels::new()
+            .with_destination(workload)
+            .with_source(src_workload);
+        let workload_uid = workload.uid.to_owned();
+        let hostname = workload.hostname.to_owned();
+        let rdns = match state.get_ips_for_hostname(&workload.hostname) {
+            Some(r) => {
+                metrics
+                    .as_ref()
+                    .on_demand_dns_cache_hits
+                    .get_or_create(&labels)
+                    .inc();
+                r
+            }
+            None => {
+                metrics
+                    .as_ref()
+                    .on_demand_dns_cache_misses
+                    .get_or_create(&labels)
+                    .inc();
+                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                // try to get it again
+                let updated_rdns = state.get_ips_for_hostname(&hostname);
+                match updated_rdns {
+                    Some(rdns) => rdns,
+                    None => {
+                        return Err(Error::NoResolvedAddresses(workload_uid));
+                    }
+                }
+            }
+        };
+
+        // TODO: add more sophisticated routing logic, perhaps based on ipv4/ipv6 support underneath us.
+        // if/when we support that, this function may need to move to get access to the necessary metadata.
+        // Randomly pick an IP
+        // TODO: do this more efficiently, and not just randomly
+        let Some(ip) = rdns.ips.iter().choose(&mut rand::thread_rng()) else {
+            return Err(Error::EmptyResolvedAddresses(workload_uid));
+        };
+        Ok(*ip)
+    }
+
+    async fn resolve_on_demand_dns(mut state: DemandProxyState, workload: &Workload) {
+        let workload_uid = workload.uid.to_owned();
+        let hostname = workload.hostname.to_owned();
+        async move {
```

Comment:
I may be missing something but it seems like we don't need a new async block since the function is async now.

Reply:
good call, thanks https://github.com/istio/ztunnel/pull/612/commits/54307a9281c4a54a26ca43aaad1d1be3e7c34282